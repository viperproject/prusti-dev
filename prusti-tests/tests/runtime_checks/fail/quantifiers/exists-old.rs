//@run: 101
use prusti_contracts::*;

#[derive(Clone)]
struct VecWrapper {
    content: Vec<i32>,
}

impl VecWrapper {
    #[trusted]
    #[insert_runtime_check]
    #[ensures(result.len() == 8)]
    pub fn new() -> Self {
        Self {
            content: vec![1,2,3,4,5,6,7,8],
        }
    }

    #[pure]
    #[trusted]
    #[insert_runtime_check]
    #[requires(i < self.len())]
    pub fn lookup(&self, i: usize) -> i32 {
        *self.content.get(i).unwrap()
    }

    #[pure]
    #[trusted]
    pub fn len(&self) -> usize {
        self.content.len()
    }

    #[trusted]
    #[insert_runtime_check]
    #[requires(i < self.len())]
    #[insert_runtime_check]
    #[ensures(old(self.len()) == self.len())]
    // failing quantifier: all elements stay untouched
    #[insert_runtime_check]
    #[ensures(forall(|j: usize| (j < self.len()) ==> self.lookup(j) == old(self.lookup(j))))]
    pub fn set(&mut self, i: usize, x: i32) {
        self.content[i] = x;
    }
}

#[trusted]
fn main() {
    let mut vec = VecWrapper::new();
    vec.set(4, 55);
    println!("executed");

}

