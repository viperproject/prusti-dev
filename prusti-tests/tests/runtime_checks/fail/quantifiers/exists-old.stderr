thread 'main' panicked at $DIR/exists-old.rs:41:5:
Prusti Runtime Checks: Contract #[ensures(forall(|j: usize| (j < self.len()) ==> self.lookup(j) == old(self.lookup(j))))] was violated at runtime
	> expression (j < self.len()) ==> self.lookup(j) == old(self.lookup(j)) was violated for index j=4
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
